[ci skip] Add composite primary key release note (#49179)
rails/rails: guides/source/7_1_release_notes.md

### Composite primary keys

> Applications can now create database tables with composite primary keys, and Rails will be able to derive these keys directly from the schema. This feature is particularly beneficial for many-to-many relationships and other complex data models where a single column is insufficient to uniquely identify a record.

Reviewer: Rails was previously able to create database tables with CPKs, right? 🤔 What do you think about highlighting that it's a change in CPK support at the model level, e.g.
```suggestion
Composite primary keys are now supported at both the database and application level. Rails is able to derive these keys directly from the schema. This feature is particularly beneficial for many-to-many relationships and other complex data models where a single column is insufficient to uniquely identify a record.
```

Author: True, I like your suggestion, will change it

> Applications can now create database tables with composite primary keys, and Rails will be able to derive these keys directly from the schema. This feature is particularly beneficial for many-to-many relationships and other complex data models where a single column is insufficient to uniquely identify a record.

Reviewer 2: It would be great if could also mention a primary PR for the reference.

> Active Record has been updated to fully support composite primary keys. All parts of the composite primary key will be included in the SQL generated by the `#reload`, `#update`, and `#delete methods`. Methods like `#first` and `#last` will use the full composite primary key in the `ORDER BY` statement.

Reviewer: Suggestion to condense this a bit:
```suggestion
The SQL generated by query methods in Active Record (e.g. `#reload`, `#update`, `#delete`) will contain all parts of the composite primary key. Methods like `#first` and `#last` will use the full composite primary key in `ORDER BY` statements.
```

> The `query_constraints`` macro can be used as a "virtual primary key" to achieve the same behaviors without modifying the database schema.

Reviewer: ```suggestion
The `query_constraints` macro can be used as a "virtual primary key" to achieve the same behavior without modifying the database schema.
```

> Similarly, the `query_constraints:` option is accepted by associations to serve as a composite foreign key. This option configures the list of columns used for accessing the associated record.

Reviewer: Tiny nit: I'd use active instead of passive voice here to make this more direct
```suggestion
Similarly, associations accept a `query_constraints:` option. This option serves as a composite foreign key, configuring the list of columns used for accessing the associated record.
```

> How to use:

Reviewer: I would probably opt to take out this `How to use` section (let's move it to the guides?), and instead to provide a simple example for the `query_constraints` macro and a simple example of using `query_constraints` as an option on an association.

Reviewer 3: If we *do* add a composite primary keys guide or large page, that would be an obvious place to point to.

Author: Done! Changed this big section in favor of a few small ones

```ruby
    create_table :travel_routes, primary_key: [:origin, :destination], force: true do |t|
      t.string :origin
      t.string :destination
      t.text :description
    end

    create_table :travel_route_reviews, primary_key: [:origin, :destination], force: true do |t|
      t.string :travel_route_origin
      t.string :travel_route_destination
      t.text :body
      t.integer :author_id
    end

    class TravelRoute < ActiveRecord::Base; end

    class TravelRouteReview < ActiveRecord::Base
      belongs_to :travel_route, query_constraints: [:travel_route_origin, :travel_route_destination]
    end

    travel_route = TravelRoute.create(origin: "Ottawa", destination: "Montreal")
    travel_route.id # => ["Ottawa", "Paris"]

    travel_route.update(description: "Ottawa to Montreal route")
    # UPDATE travel_routes SET description = 'Ottawa to Montreal route'
    #  WHERE origin = 'Ottawa' AND destination = 'Montreal'

    review = TravelRouteReview.new(author_id: 1, body: "I often take this route!")
    review.travel_route = travel_route
    review.travel_route_origin # => "Ottawa"
    review.travel_route_destination # => "Montreal"

    travel_route.destroy

>     # DELETE FROM travel_routes WHERE origin = 'Ottawa' and destination = 'Montreal'

Reviewer 2: ```suggestion
    # DELETE FROM travel_routes WHERE origin = 'Ottawa' AND destination = 'Montreal'
```

> ```

Reviewer 2: Fix the indentation for the ruby block

Author: Thanks, we decided to remove this section from the notes but I removed the indentation from the new ones
